### PR TITLE
Add controller to generate addon label on cluster

### PR DESCRIPTION
**Pull request by elgnay**
The related issue: https://github.com/open-cluster-management/backlog/issues/15279
Enhancement proposal: https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/17-cluster-features
Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/spoke/addon/discovery_controller.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		// addon is deleting
+		if !addOn.DeletionTimestamp.IsZero() {
+			continue
+		}
+		key := addOnFeaturePrefix + addOn.Name
+		newLabels[key] = getAddOnLabelValue(addOn)
+	}
+
+	// no work if the cluster labels have no change
+	if (len(newLabels) == 0 && len(cluster.Labels) == 0) || reflect.DeepEqual(newLabels, cluster.Labels) {
```
Comment (qiujian16) [inline pkg/spoke/addon/discovery_controller.go]: do we need to compare length here?

File: pkg/spoke/addon/discovery_controller.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+	return factory.New().
+		WithInformers(hubClusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			hubAddOnInformers.Informer()).
+		WithSync(c.sync).
```
Comment (qiujian16) [inline pkg/spoke/addon/discovery_controller.go]: I think you would need a resync here to handle unexpected deletion?

File: pkg/spoke/addon/discovery_controller.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+func getAddOnLabelValue(addOn *addonv1alpha1.ManagedClusterAddOn) string {
+	availableCondition := meta.FindStatusCondition(addOn.Status.Conditions, addonv1alpha1.ManagedClusterAddOnConditionAvailable)
+	if availableCondition == nil {
+		return "unreachable"
+	}
+
+	switch availableCondition.Status {
+	case metav1.ConditionTrue:
+		return "available"
```
Comment (qiujian16) [inline pkg/spoke/addon/discovery_controller.go]: make it const

File: pkg/hub/addon/discovery_controller.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	c := &addOnFeatureDiscoveryController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		addOnLister:   addOnInformers.Lister(),
+		recorder:      recorder,
+	}
+
+	return factory.New().
+		WithInformers(clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go:60]: two cluster informers?

File: pkg/hub/addon/discovery_controller.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	if len(namespace) > 0 {
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: it needs doc here, otherwise it is really strange

File: pkg/hub/addon/discovery_controller.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		// sync a particular addon
+		return c.syncAddOn(ctx, namespace, name)
+	}
+
+	// sync all labels of the managed cluster
+	clusterName := name
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: this seems redundant

File: pkg/hub/addon/discovery_controller.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		// addon is deleted
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
+	}
+	if err != nil {
+		return err
+	}
+
+	// addon is deleting
+	if !addOn.DeletionTimestamp.IsZero() {
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: it seems to me that using https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourcemerge/object_merger.go#L138 can be much easier.

File: pkg/spoke/addon/discovery_controller.go (hunk 1 above)
Comment (elgnay) [reply pkg/spoke/addon/discovery_controller.go]: Yes. An empty map is not equal to nil, that will result in unnecessary update.

File: pkg/spoke/addon/discovery_controller.go (hunk 2 above)
Comment (elgnay) [reply pkg/spoke/addon/discovery_controller.go]: Added.

File: pkg/spoke/addon/discovery_controller.go (hunk 3 above)
Comment (elgnay) [reply pkg/spoke/addon/discovery_controller.go]: Done.

File: pkg/hub/addon/discovery_controller.go (hunk 4 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go:60]: Removed the redundant one.

File: pkg/hub/addon/discovery_controller.go (hunk 5 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go]: Done.

File: pkg/hub/addon/discovery_controller.go (hunk 6 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go]: Removed.

File: pkg/hub/addon/discovery_controller.go (hunk 7 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go]: Code updated.

File: pkg/hub/addon/discovery_controller.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+		// addon is deleting
+		if !addOn.DeletionTimestamp.IsZero() {
+			continue
+		}
+		key := addOnFeaturePrefix + addOn.Name
+		newLabels[key] = getAddOnLabelValue(addOn)
+	}
+
+	// no work if the cluster labels have no change
+	if (len(newLabels) == 0 && len(cluster.Labels) == 0) || reflect.DeepEqual(newLabels, cluster.Labels) {
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: can we use MergeMap here also?

File: pkg/hub/addon/discovery_controller.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+		syncCtx.Queue().Add(cluster.Name)
+	}
+	return nil
+}
+
+func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, clusterName, addOnName string) error {
+	klog.V(4).Infof("Reconciling addOn %q", addOnName)
+
+	addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
```
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: I might get cluster here, merge labels, then update cluster. So we do not need to differentiate deleteAddonLabel and updateAddonLabel.

File: pkg/hub/addon/discovery_controller.go (hunk 8 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go]: Yes. The code updated.

File: pkg/hub/addon/discovery_controller.go (hunk 9 above)
Comment (elgnay) [reply pkg/hub/addon/discovery_controller.go]: Code updated as suggested.

File: pkg/hub/addon/discovery_controller.go (hunk 8 above)
Comment (qiujian16) [inline pkg/hub/addon/discovery_controller.go]: nit: if you  use apiequality.sementic.DeepEqual you would not need to compare the length

File: pkg/hub/addon/discovery_controller.go (hunk 9 above)
Comment (qiujian16) [reply pkg/hub/addon/discovery_controller.go]: what i  meant is

```go

labels := map[string]string{}
addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
switch {
case errors.IsNotFound(err):
  labels[addonName + "-"] = ""
case err != nil:
  return err
case !addOn.DeletionTimestamp.IsZero():
  labels[addonName + "-"] = ""
default:
  labels[addonName] = getAddOnLabelValue(addOn)
}

cluster, err := c.clusterLister.Get(clusterName)
if err != nil {
	return fmt.Errorf("unable to find cluster with name %q: %w", clusterName, err)
}

resourcemerge.MergeMap(&modified, &cluster.Labels,  labels)

if modified {
 // update things
}

```

File: pkg/hub/addon/discovery_controller.go (hunk 8 above)
Comment (qiujian16) [reply pkg/hub/addon/discovery_controller.go]: I might do like:

```go
newLabels := map[string]string{}
addOns, err := c.addOnLister.ManagedClusterAddOns(clusterName).List(labels.Everything())
	if err != nil {
		return fmt.Errorf("unable to list addOns of cluster %q: %w", clusterName, err)
	}
	for _, addOn := range addOns {
		// addon is deleting
		if !addOn.DeletionTimestamp.IsZero() {
			continue
		}
		key := addOnFeaturePrefix + addOn.Name
		newLabels[key] = getAddOnLabelValue(addOn)
	}

newCluster := cluster.DeepCopy()
for key, value := range newCluster.Labels {
		if !strings.HasPrefix(key, addOnFeaturePrefix) {
			continue
		}
              if _, ok := newLabels[key]; !ok {
                    newLabels[key + "-"] = value
              }
		
	}

resourcemerge.MergeMap(&modified, &newCluster.Labels,  newLabels)

if modified {
// update things
}

```

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**